vhost-device-spi/src/vhu_spi.rs

```diff
@@ -0,0 +1,948 @@
+                    }
+
+                    (tx_buf, rx_buf, len)
+                }
+
+                4 => {
+                    // full-duplex transfer
+                    let tx_desc_buf = descriptors[1];
+                    let rx_desc_buf = descriptors[2];
+
+                    let tx_len = tx_desc_buf.len();
+                    let rx_len = rx_desc_buf.len();
+
+                    if tx_len == 0 {
+                        return Err(Error::UnexpectedDescriptorSize(1, tx_len));
+                    }
+
+                    if rx_len == 0 {
+                        return Err(Error::UnexpectedDescriptorSize(1, rx_len));
+                    }
+
+                    if tx_len != rx_len {
+                        return Err(Error::TxRxTrnasLenNotEqual(tx_len, rx_len));
+                    }
+
+                    rx_buf = vec![0; rx_len as usize];
+                    tx_buf = vec![0; tx_len as usize];
+
+                    desc_chain
+                        .memory()
+                        .read(&mut tx_buf, tx_desc_buf.addr())
+                        .map_err(|_| Error::DescriptorReadFailed)?;
+
+                    (tx_buf, rx_buf, tx_len)
+                }
+
+                _ => {
+                    return Err(Error::UnexpectedDescriptorCount(descriptors.len()));
+                }
+            };
+
+            reqs.push(SpiTransReq {
+                tx_buf,
+                rx_buf,
+                trans_len,
+                speed_hz: out_hdr.freq.to_native(),
+                delay_usecs: (out_hdr.cs_delay_hold_ns.to_native() / 1000) as u16,
+                bits_per_word: out_hdr.bits_per_word,
+                cs_change: out_hdr.cs_change,
+                tx_nbits: out_hdr.tx_nbits,
+                rx_nbits: out_hdr.rx_nbits,
+                word_delay_usecs: (out_hdr.word_delay_ns.to_native() / 1000) as u8,
+                mode: out_hdr.mode.to_native(),
+                cs_id: out_hdr.chip_select_id,
+            });
+        }
+
+        let in_hdr = match self.spi_ctrl.check_trans_params(&mut reqs[0]) {
+            true => VirtioSpiTransferResult {
+                status: match self.spi_ctrl.transfer(&mut reqs) {
+                    Ok(()) => VIRTIO_SPI_TRANS_OK,
+                    Err(_) => VIRTIO_SPI_TRANS_ERR,
+                },
+            },
+            _ => VirtioSpiTransferResult {
+                status: VIRTIO_SPI_PARAM_ERR,
+            },
+        };
+
+        for (i, desc_chain) in requests.iter().enumerate() {
+            let descriptors: Vec<_> = desc_chain.clone().collect();
+            let desc_in_hdr = descriptors[descriptors.len() - 1];
+            let mut len = size_of::<VirtioSpiTransferResult>() as u32;
+
+            if descriptors.len() == 3 {
+                let desc_buf = descriptors[1];
+
+                if desc_buf.is_write_only() {
+                    desc_chain
+                        .memory()
+                        .write(&reqs[i].rx_buf, desc_buf.addr())
```

Comment:
How do you know that `rx_buff` has been already populated at this point?

Reply:
vhost-device-spi uses ioctl to implement the transfer process, the same as vhost i2c.

In Linux, at the end of spidev_message function, there is a step to copy rx data out of bounce buffer, as the following code shows:

        for (n = n_xfers, k_tmp = k_xfers, u_tmp = u_xfers;
                        n;
                        n--, k_tmp++, u_tmp++) {
                if (u_tmp->rx_buf) {
                        if (copy_to_user((u8 __user *)
                                        (uintptr_t) u_tmp->rx_buf, k_tmp->rx_buf,
                                        u_tmp->len)) {
                                status = -EFAULT;
                                goto done;
                        }
                }
        }

So if the ioctl finished, we can know that the rx data is already copied to the user level rx_buf.

Reply:
Resolved.